### PR TITLE
expression: implement vectorized evaluation for builtinCastJSONAsTimeSig

### PR DESCRIPTION
**Pull request by ekalinin**
PCP #12101

### What problem does this PR solve? 

This PR implements vectorized `builtinCastJSONAsTimeSig`.

### What is changed and how it works?

```sh
➜ make vectorized-bench VB_FILE=Cast VB_FUNC=builtinCastJSONAsTimeSig
cd ./expression && \
	go test -v -benchmem \
		-bench=BenchmarkVectorizedBuiltinCastFunc \
		-run=BenchmarkVectorizedBuiltinCastFunc \
		-args "builtinCastJSONAsTimeSig"
goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinCastFunc/builtinCastJSONAsTimeSig-VecBuiltinFunc-12         	    3174	   1370976 ns/op	  147458 B/op	    7168 allocs/op
BenchmarkVectorizedBuiltinCastFunc/builtinCastJSONAsTimeSig-NonVecBuiltinFunc-12      	     745	   1617424 ns/op	  147457 B/op	    7168 allocs/op
PASS
ok  	github.com/pingcap/tidb/expression	5.793s
```

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: expression/builtin_cast_vec_test.go [hunk 1]
```diff
@@ -70,6 +70,7 @@ var vecBuiltinCastCases = map[string][]vecExprBenchCase{
 				&timeStrGener{},
 				&dataStrGener{},
 			}},
+		{retEvalType: types.ETDatetime, childrenTypes: []types.EvalType{types.ETJson}},
```
Comment (wshwsh12) [inline expression/builtin_cast_vec_test.go]: Should we generate some jsons that can cast to valid time ?

Comment (ekalinin) [reply expression/builtin_cast_vec_test.go]: Added `jsonTimeGener`.

**Comment by ekalinin** [comment]
As I see #13316 was merged :( So my PR becomes a duplicate :(

**Comment by SunRunAway** [comment]
I think the test case is still needed. :)

**Comment by ekalinin** [comment]
Glad it still useful.

File: expression/builtin_cast_vec_test.go [hunk 2]
```diff
@@ -35,7 +35,6 @@ var vecBuiltinCastCases = map[string][]vecExprBenchCase{
 		{retEvalType: types.ETReal, childrenTypes: []types.EvalType{types.ETInt}},
 		{retEvalType: types.ETDuration, childrenTypes: []types.EvalType{types.ETInt}, geners: []dataGenerator{new(randDurInt)}},
 		{retEvalType: types.ETReal, childrenTypes: []types.EvalType{types.ETReal}},
-		{retEvalType: types.ETReal, childrenTypes: []types.EvalType{types.ETJson}},
```
Comment (wshwsh12) [inline expression/builtin_cast_vec_test.go]: Why remove this test case?

Comment (ekalinin) [reply expression/builtin_cast_vec_test.go]: Ooops. My bad. Reverted.

**Pull request merged**